src/main/java/org/opentripplanner/middleware/triptracker/ManageLegTraversal.java

```diff
@@ -58,33 +58,26 @@ private static boolean canUseLeg(Leg leg) {
     }
 
     /**
-     * Get the expected leg by comparing the current time against the start and end time of each leg.
+     * Get the expected leg by comparing the current position against all points on a leg. Comparing against all points
+     * on a leg gives greater accuracy than just comparing against the start and end of a leg.
      */
     @Nullable
-    public static Leg getExpectedLeg(Instant timeNow, Itinerary itinerary) {
+    public static Leg getExpectedLeg(Coordinates position, Itinerary itinerary) {
+        double shortestDistance = Double.MAX_VALUE;
+        Leg nearestLeg = null;
         if (canUseTripLegs(itinerary)) {
             for (int i = 0; i < itinerary.legs.size(); i++) {
                 Leg leg = itinerary.legs.get(i);
-                if (i == 0 &&
-                    leg.mode.equalsIgnoreCase("walk") &&
-                    leg.startTime != null &&
-                    timeNow.isBefore(leg.startTime.toInstant())) {
-                    // If the first leg is a walk leg and the traveler has decided to start the trip early.
-                    return leg;
-                }
-                if (leg.startTime != null &&
-                    leg.endTime != null &&
-                    isTimeInRange(
-                        leg.startTime.toInstant(),
-                        // Offset the end time by a faction to avoid exact times being attributed to the wrong leg.
-                        leg.endTime.toInstant().minus(1, ChronoUnit.MILLIS),
-                        timeNow
-                    )) {
-                    return leg;
+                for (Coordinates positionOnLeg : getAllLegPositions(leg)) {
```

Comment:
Sorry for going back and forth, but the snippet you had previously is still useful when someone engages the next leg after taking a turn as instructed. In that case we want to not show the turn instruction.
So how about using `isPointBetween` first, then fallback on `getDistance`?

```java
                List<Coordinates> allPositions = getAllLegPositions(leg);
                for (int j = 0; j < allPositions.size() - 1; j++) {
                    if (isPointBetween(allPositions.get(j), allPositions.get(j + 1), position)) {
                        return leg;
```

Reply:
No worries. Updated in: https://github.com/ibi-group/otp-middleware/pull/227/commits/e0e990c067a831c06664bd93009fc4582463d00c